html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js

```diff
@@ -20,3 +20,67 @@ structuredCloneBatteryOfTests.push({
 });
 
 // TODO: ImageBitmap
+
+structuredCloneBatteryOfTests.push({
+  description: 'A detached ArrayBuffer cannot be transferred',
+  async f(runner, t) {
+    const buffer = new ArrayBuffer();
+    await runner.structuredClone(buffer, [buffer]);
+    await promise_rejects_dom(
+      t,
+      "DataCloneError",
+      runner.structuredClone(buffer, [buffer])
+    );
+  }
+});
+
+structuredCloneBatteryOfTests.push({
+  description: 'A detached platform object cannot be transferred',
+  async f(runner, t) {
+    const {port1} = new MessageChannel();
+    await runner.structuredClone(port1, [port1]);
+    await promise_rejects_dom(
+      t,
+      "DataCloneError",
+      runner.structuredClone(port1, [port1])
+    );
+  }
+});
+
+structuredCloneBatteryOfTests.push({
+  description: 'Transferring a non-transferable platform object fails',
+  async f(runner, t) {
+    const blob = new Blob();
+    await promise_rejects_dom(
+      t,
+      "DataCloneError",
+      runner.structuredClone(blob, [blob])
+    );
+  }
+});
+
+structuredCloneBatteryOfTests.push({
+  description: 'An object whose interface is deleted from the global object must still be received',
+  async f(runner) {
+    const {port1} = new MessageChannel();
+    const messagePortInterface = globalThis.MessagePort;
+    delete globalThis.MessagePort;
+    try {
+      const transfer = await runner.structuredClone(port1, [port1]);
+      assert_true(transfer instanceof messagePortInterface);
+    } finally {
+      globalThis.MessagePort = messagePortInterface;
+    }
+  }
+});
+
+structuredCloneBatteryOfTests.push({
+  description: 'A subclass instance will be received as its closest transferable superclass',
+  async f(runner) {
+    // MessagePort doesn't have a constructor, so we must use something else.
+    class OffscreenCanvasSubclass extends OffscreenCanvas {}
```

Comment:
OffscreenCanvas isn't implemented in Firefox and Safari. Can we use something else? I think ImageBitmap would be good.

Reply:
`ImageBitmap`, just like `MessagePort`, doesn't have a constructor, so you can't instantiate a subclass. You could create a new object that has `ImageBitmap.prototype` in its prototype chain, or you could change the prototype of an existing instance – which probably should be added as tests of their own – but neither is what I'm testing here.

The only constructable transferable interface that's implemented in every browser seems to be `ReadableStream`, and it's only transferable in Chrome.

Reply:
Well, hmm. I would slightly prefer ReadableStream if that still gets the same coverage, as transferability seems likely to be implemented sooner than OffscreenCanvas.

In either event, it sounds like this test should be made optional using https://web-platform-tests.org/writing-tests/testharness-api.html#optional-features